File: backend/internal/handler/media/media.go
package handler

import (
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	"github.com/spacelab/backend/internal/config"
	"github.com/spacelab/backend/internal/model"
	"github.com/spacelab/backend/internal/utils"
	"gorm.io/gorm"
)

type MediaHandler struct {
	cfg *config.Config
	db  *gorm.DB
}

func NewMediaHandler(cfg *config.Config, db *gorm.DB) *MediaHandler {
	return &MediaHandler{cfg: cfg, db: db}
}

// Upload 上传文件
func (h *MediaHandler) Upload(c *gin.Context) {
	file, err := c.FormFile("file")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	// 检查文件大小
	if file.Size > h.cfg.MaxUploadSize {
		c.JSON(http.StatusBadRequest, gin.H{"error": fmt.Sprintf("File size exceeds maximum allowed size (%d bytes)", h.cfg.MaxUploadSize)})
		return
	}

	// 验证文件类型
	validTypes := map[string]bool{
		"image/jpeg": true,
		"image/png":  true,
		"image/gif":  true,
		"image/webp": true,
		"video/mp4":  true,
		"video/webm": true,
	}

	if !validTypes[file.Header.Get("Content-Type")] {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid file type"})
		return
	}

	// 创建上传目录
	uploadDir := filepath.Join(h.cfg.UploadPath, time.Now().Format("2006/01"))
	if err := os.MkdirAll(uploadDir, 0755); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create upload directory"})
		return
	}

	// S10: 白名单验证扩展名
	validExts := map[string]bool{
		".jpg": true, ".jpeg": true, ".png": true, ".gif": true, ".webp": true,
		".mp4": true, ".webm": true,
	}
	ext := strings.ToLower(filepath.Ext(file.Filename))
	if !validExts[ext] {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid file extension"})
		return
	}

	// 路径遍历防护：确保文件名不包含路径分隔符或 ..
	cleanName := filepath.Base(file.Filename)
	if cleanName != file.Filename || strings.Contains(file.Filename, "..") {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid filename"})
		return
	}

	// L6: 使用临时文件写入，成功后重命名
	tempFile, err := os.CreateTemp(uploadDir, "upload-*"+ext)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create temp file"})
		return
	}
	tempPath := tempFile.Name()

	// 保存文件到临时文件
	src, err := file.Open()
	if err != nil {
		tempFile.Close()
		os.Remove(tempPath)
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	defer src.Close()

	if _, err = tempFile.ReadFrom(src); err != nil {
		tempFile.Close()
		os.Remove(tempPath)
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	tempFile.Close()

	// Magic bytes 验证：读取文件头确认真实类型
	detectedType, err := detectFileType(tempPath)
	if err != nil || !validTypes[detectedType] {
		os.Remove(tempPath)
		c.JSON(http.StatusBadRequest, gin.H{"error": "File content does not match allowed types"})
		return
	}

	// 生成最终文件名（UUID + 白名单扩展名）
	filename := fmt.Sprintf("%s%s", uuid.New().String(), ext)
	finalPath := filepath.Join(uploadDir, filename)
	if err := os.Rename(tempPath, finalPath); err != nil {
		os.Remove(tempPath)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to save file"})
		return
	}

	// 获取文件信息
	info, err := os.Stat(finalPath)
	if err != nil {
		os.Remove(finalPath)
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	// 存储到数据库
	uploaderID, _ := uuid.Parse(c.GetString("user_id"))
	mediaAsset := model.MediaAsset{
		ID:           uuid.New(),
		Filename:     filename,
		OriginalName: file.Filename,
		StoragePath:  finalPath,
		MimeType:     detectedType, // 使用 magic bytes 检测的真实类型，而非客户端 Content-Type
		Size:         info.Size(),
		Type:         getMediaType(detectedType),
		UploadedBy:   uploaderID,
		CreatedAt:    time.Now(),
	}

	if err := h.db.Create(&mediaAsset).Error; err != nil {
		os.Remove(finalPath) // 回滚：删除文件
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	// 返回文件 URL（使用环境变量配置基础 URL）
	baseURL := utils.GetEnv("API_BASE_URL", fmt.Sprintf("http://localhost:%d", h.cfg.ServerPort))
	fileURL := fmt.Sprintf("%s/uploads/%s", baseURL, filepath.ToSlash(filepath.Join(filepath.Dir(filename), filename)))

	c.JSON(http.StatusCreated, gin.H{
		"id":        mediaAsset.ID.String(),
		"url":       fileURL,
		"name":      mediaAsset.OriginalName,
		"type":      mediaAsset.Type,
		"size":      mediaAsset.Size,
		"mime_type": mediaAsset.MimeType,
	})
}

// List 获取媒体列表
func (h *MediaHandler) List(c *gin.Context) {
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("page_size", "20"))
	fileType := c.Query("type")

	if page < 1 {
		page = 1
	}

	var mediaAssets []model.MediaAsset
	var total int64

	query := h.db.Model(&model.MediaAsset{})

	if fileType != "" {
		query = query.Where("type = ?", fileType)
	}

	query.Count(&total)
	offset := (page - 1) * pageSize
	result := query.Offset(offset).Limit(pageSize).Order("created_at DESC").Find(&mediaAssets)

	if result.Error != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": result.Error.Error()})
		return
	}

	response := struct {
		Assets     []model.MediaAsset `json:"assets"`
		Total      int64              `json:"total"`
		Page       int                `json:"page"`
		PageSize   int                `json:"page_size"`
		TotalPages int                `json:"total_pages"`
	}{
		Assets:     mediaAssets,
		Total:      total,
		Page:       page,
		PageSize:   pageSize,
		TotalPages: int((total + int64(pageSize-1)) / int64(pageSize)),
	}

	c.JSON(http.StatusOK, response)
}

// Delete 删除媒体文件
func (h *MediaHandler) Delete(c *gin.Context) {
	id := c.Param("id")

	var mediaAsset model.MediaAsset
	result := h.db.Where("id = ?", id).First(&mediaAsset)

	if result.Error != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Media not found"})
		return
	}

	// 权限检查：只有 admin 或资源所有者可删除
	currentUserID := c.GetString("user_id")
	currentRole := c.GetString("role")
	if mediaAsset.UploadedBy != uuid.Nil && mediaAsset.UploadedBy.String() != currentUserID && currentRole != "admin" {
		c.JSON(http.StatusForbidden, gin.H{"error": "Permission denied"})
		return
	}

	// 路径遍历防护：确保 StoragePath 在上传目录内
	absPath, err := filepath.Abs(mediaAsset.StoragePath)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Invalid file path"})
		return
	}
	absUpload, err := filepath.Abs(h.cfg.UploadPath)
	if err != nil || !strings.HasPrefix(absPath, absUpload) {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid storage path"})
		return
	}

	// 删除物理文件
	if err := os.Remove(mediaAsset.StoragePath); err != nil && !os.IsNotExist(err) {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete file"})
		return
	}

	// 删除数据库记录
	if err := h.db.Delete(&mediaAsset).Error; err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to delete from database"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Media deleted successfully"})
}

// Get 获取媒体详情
func (h *MediaHandler) Get(c *gin.Context) {
	id := c.Param("id")

	var mediaAsset model.MediaAsset
	result := h.db.Where("id = ?", id).First(&mediaAsset)

	if result.Error != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Media not found"})
		return
	}

	c.JSON(http.StatusOK, mediaAsset)
}

func getMediaType(mimeType string) string {
	if mimeType == "" {
		return "unknown"
	}

	if len(mimeType) > 5 && mimeType[:5] == "image" {
		return "image"
	}
	if len(mimeType) > 5 && mimeType[:5] == "video" {
		return "video"
	}

	return "unknown"
}

// detectFileType 通过读取文件头部 magic bytes 判断真实 MIME 类型
// 防止攻击者伪造 Content-Type / 扩展名上传恶意文件
func detectFileType(path string) (string, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer f.Close()

	buf := make([]byte, 32)
	n, err := f.Read(buf)
	if err != nil || n < 4 {
		return "", fmt.Errorf("cannot read file header")
	}

	switch {
	case len(buf) >= 3 && buf[0] == 0xFF && buf[1] == 0xD8 && buf[2] == 0xFF:
		return "image/jpeg", nil
	case len(buf) >= 8 && string(buf[:8]) == "\x89PNG\r\n\x1a\n":
		return "image/png", nil
	case len(buf) >= 6 && (string(buf[:6]) == "GIF87a" || string(buf[:6]) == "GIF89a"):
		return "image/gif", nil
	case len(buf) >= 4 && string(buf[:4]) == "RIFF" && len(buf) >= 12 && string(buf[8:12]) == "WEBP":
		return "image/webp", nil
	case len(buf) >= 8 && (string(buf[:4]) == "ftyp" || (len(buf) >= 12 && string(buf[4:8]) == "ftyp")):
		// MP4 / MOV 等容器格式均以 ftyp box 开头
		return "video/mp4", nil
	case len(buf) >= 4 && string(buf[:4]) == "\x1a\x45\xdf\xa3":
		return "video/webm", nil
	default:
		return "", fmt.Errorf("unknown file type")
	}
}

File: cf-worker/src/routes/media.ts
/**
 * 媒体上传路由 — 对齐 Go 版 handler/media/media.go
 */
import { Hono } from "hono";
import { eq, and, sql, desc } from "drizzle-orm";
import { drizzle } from "drizzle-orm/d1";
import type { AuthVariables } from "../middleware/auth";
import { authMiddleware } from "../middleware/auth";
import { mediaAssets } from "../db/schema";
import { generateUUID } from "../utils/uuid";
import { nowISO } from "../utils/time";

type AppContext = { Bindings: Env; Variables: AuthVariables };

const mediaRouter = new Hono<AppContext>();

// ── 常量 ──────────────────────────────────────────────────────────

const MAX_UPLOAD_SIZE = 10 * 1024 * 1024; // 10MB

const VALID_MIME_TYPES: Record<string, boolean> = {
  "image/jpeg": true,
  "image/png": true,
  "image/gif": true,
  "image/webp": true,
  "video/mp4": true,
  "video/webm": true,
};

const VALID_EXTENSIONS: Record<string, boolean> = {
  ".jpg": true,
  ".jpeg": true,
  ".png": true,
  ".gif": true,
  ".webp": true,
  ".mp4": true,
  ".webm": true,
};

// ── 工具函数 ──────────────────────────────────────────────────────

/** 统一错误处理 */
function handleError(c: any, err: unknown) {
  if (err && typeof err === "object" && "status" in err && "message" in err) {
    const e = err as { status: number; message: string };
    return c.json({ error: e.message }, e.status);
  }
  console.error("Unexpected error:", err);
  return c.json({ error: "Internal Server Error" }, 500);
}

/** 获取 R2 公开访问基础 URL */
function getR2PublicUrl(env: Env): string {
  return env.R2_PUBLIC_URL || "https://pub-xxx.r2.dev";
}

/**
 * 通过 magic bytes 检测文件真实 MIME 类型
 */
function detectFileType(bytes: Uint8Array): string | null {
  if (bytes.length < 4) return null;

  // JPEG: FF D8 FF
  if (bytes[0] === 0xff && bytes[1] === 0xd8 && bytes[2] === 0xff) {
    return "image/jpeg";
  }

  // PNG: 89 50 4E 47 0D 0A 1A 0A
  if (
    bytes.length >= 8 &&
    bytes[0] === 0x89 &&
    bytes[1] === 0x50 &&
    bytes[2] === 0x4e &&
    bytes[3] === 0x47 &&
    bytes[4] === 0x0d &&
    bytes[5] === 0x0a &&
    bytes[6] === 0x1a &&
    bytes[7] === 0x0a
  ) {
    return "image/png";
  }

  // GIF: GIF87a or GIF89a
  if (
    bytes.length >= 6 &&
    bytes[0] === 0x47 &&
    bytes[1] === 0x49 &&
    bytes[2] === 0x46 &&
    bytes[3] === 0x38 &&
    (bytes[4] === 0x37 || bytes[4] === 0x39) &&
    bytes[5] === 0x61
  ) {
    return "image/gif";
  }

  // WEBP: RIFF....WEBP
  if (
    bytes.length >= 12 &&
    bytes[0] === 0x52 &&
    bytes[1] === 0x49 &&
    bytes[2] === 0x46 &&
    bytes[3] === 0x46 &&
    bytes[8] === 0x57 &&
    bytes[9] === 0x45 &&
    bytes[10] === 0x42 &&
    bytes[11] === 0x50
  ) {
    return "image/webp";
  }

  // MP4: ftyp box
  if (
    bytes.length >= 8 &&
    ((bytes[4] === 0x66 && bytes[5] === 0x74 && bytes[6] === 0x79 && bytes[7] === 0x70) ||
      (bytes.length >= 12 &&
        bytes[0] === 0x66 &&
        bytes[1] === 0x74 &&
        bytes[2] === 0x79 &&
        bytes[3] === 0x70))
  ) {
    return "video/mp4";
  }

  // WebM: 1A 45 DF A3
  if (
    bytes.length >= 4 &&
    bytes[0] === 0x1a &&
    bytes[1] === 0x45 &&
    bytes[2] === 0xdf &&
    bytes[3] === 0xa3
  ) {
    return "video/webm";
  }

  return null;
}

/**
 * 根据 MIME 类型获取媒体类别
 */
function getMediaType(mimeType: string): string {
  if (mimeType.startsWith("image/")) {
    if (mimeType === "image/gif") return "gif";
    return "image";
  }
  if (mimeType.startsWith("video/")) return "video";
  return "unknown";
}

/**
 * 获取文件扩展名
 */
function getExtension(filename: string): string {
  const idx = filename.lastIndexOf(".");
  if (idx === -1) return "";
  return filename.slice(idx).toLowerCase();
}

// ── 公开路由 ──────────────────────────────────────────────────────

/**
 * GET / — 媒体列表
 * Go 版返回: { assets, total, page, page_size, total_pages }
 */
mediaRouter.get("/", async (c) => {
  const db = drizzle(c.env.DB);
  const page = parseInt(c.req.query("page") || "1", 10);
  const pageSize = parseInt(c.req.query("page_size") || "20", 10);
  const type = c.req.query("type");

  const conditions: any[] = [];
  if (type) {
    conditions.push(eq(mediaAssets.type, type));
  }

  const whereClause =
    conditions.length === 0
      ? undefined
      : conditions.length === 1
        ? conditions[0]
        : and(...conditions);

  const countResult = await db
    .select({ count: sql<number>`count(*)` })
    .from(mediaAssets)
    .where(whereClause);
  const total = countResult[0]?.count || 0;

  const offset = (Math.max(1, page) - 1) * pageSize;
  const assets = await db
    .select()
    .from(mediaAssets)
    .where(whereClause)
    .orderBy(desc(mediaAssets.createdAt))
    .limit(pageSize)
    .offset(offset);

  return c.json({
    assets,
    total,
    page,
    page_size: pageSize,
    total_pages: Math.ceil(total / pageSize),
  });
});

/**
 * GET /:id — 媒体详情
 * Go 版直接返回 mediaAsset 对象
 */
mediaRouter.get("/:id", async (c) => {
  const id = c.req.param("id");
  const db = drizzle(c.env.DB);

  const result = await db
    .select()
    .from(mediaAssets)
    .where(eq(mediaAssets.id, id))
    .limit(1);

  if (result.length === 0) {
    return c.json({ error: "Media not found" }, 404);
  }

  return c.json(result[0]);
});

// ── 需要认证的路由 ────────────────────────────────────────────────

/**
 * POST /upload — 文件上传
 * Go 版返回: { id, url, name, type, size, mime_type } (201)
 */
mediaRouter.post("/upload", authMiddleware, async (c) => {
  const userId = c.get("userId");

  try {
    const formData = await c.req.formData();
    const file = formData.get("file") as File | null;

    if (!file) {
      return c.json({ error: "file is required" }, 400);
    }

    // 1. 大小限制
    if (file.size > MAX_UPLOAD_SIZE) {
      return c.json(
        {
          error: `File size exceeds maximum allowed size (${MAX_UPLOAD_SIZE} bytes)`,
        },
        400
      );
    }

    // 2. 扩展名白名单
    const ext = getExtension(file.name);
    if (!VALID_EXTENSIONS[ext]) {
      return c.json({ error: "Invalid file extension" }, 400);
    }

    // 3. MIME 白名单（客户端声明）
    const clientMime = file.type;
    if (!VALID_MIME_TYPES[clientMime]) {
      return c.json({ error: "Invalid file type" }, 400);
    }

    // 4. Magic bytes 验证（读取文件头部字节验证真实类型）
    const headerBuffer = await file.slice(0, 32).arrayBuffer();
    const headerBytes = new Uint8Array(headerBuffer);
    const detectedMime = detectFileType(headerBytes);

    if (!detectedMime || !VALID_MIME_TYPES[detectedMime]) {
      return c.json(
        { error: "File content does not match allowed types" },
        400
      );
    }

    // 5. UUID 重命名
    const uuid = generateUUID();
    const filename = `${uuid}${ext}`;
    const storagePath = `media/${filename}`;

    // 6. 上传到 R2
    await c.env.MEDIA_BUCKET.put(storagePath, await file.arrayBuffer(), {
      httpMetadata: {
        contentType: detectedMime,
      },
    });

    // 7. 生成 URL（从环境变量读取 R2 公开访问地址）
    const baseUrl = getR2PublicUrl(c.env);
    const url = `${baseUrl}/${storagePath}`;

    // 8. 写入 D1 元数据
    const db = drizzle(c.env.DB);
    const now = nowISO();
    const assetId = generateUUID();

    await db.insert(mediaAssets).values({
      id: assetId,
      filename,
      originalName: file.name,
      storagePath,
      url,
      mimeType: detectedMime,
      size: file.size,
      type: getMediaType(detectedMime),
      uploadedBy: userId,
      createdAt: now,
    });

    return c.json(
      {
        id: assetId,
        url,
        name: file.name,
        type: getMediaType(detectedMime),
        size: file.size,
        mime_type: detectedMime,
      },
      201
    );
  } catch (err) {
    return handleError(c, err);
  }
});

/**
 * DELETE /:id — 文件删除
 */
mediaRouter.delete("/:id", authMiddleware, async (c) => {
  const id = c.req.param("id");
  const userId = c.get("userId");
  const role = c.get("role");
  const db = drizzle(c.env.DB);

  const result = await db
    .select()
    .from(mediaAssets)
    .where(eq(mediaAssets.id, id))
    .limit(1);

  if (result.length === 0) {
    return c.json({ error: "Media not found" }, 404);
  }

  const asset = result[0];

  // 权限检查：admin 或资源所有者
  if (asset.uploadedBy !== userId && role !== "admin") {
    return c.json({ error: "Permission denied" }, 403);
  }

  // 删除 R2 文件
  try {
    await c.env.MEDIA_BUCKET.delete(asset.storagePath);
  } catch (e) {
    console.warn("Failed to delete from R2:", e);
  }

  // 删除数据库记录
  await db.delete(mediaAssets).where(eq(mediaAssets.id, id));

  return c.json({ message: "Media deleted successfully" });
});

export { mediaRouter as mediaRoutes };
